src/plugins/mvcc_delete_plugin.cpp

```diff
@@ -0,0 +1,158 @@
+      if (chunk && chunk->get_cleanup_commit_id() == MvccData::MAX_COMMIT_ID) {
+        // Calculate metric 1 – Chunk invalidation level
+        const double invalidated_rows_ratio = static_cast<double>(chunk->invalid_row_count()) / chunk->size();
+
+        // Calculate metric 2 – Chunk Hotness
+        const CommitID lowest_end_commit_id =
+            *std::min_element(std::begin(chunk->mvcc_data()->end_cids), std::end(chunk->mvcc_data()->end_cids));
+        const CommitID commit_id_diff = TransactionManager::get().last_commit_id() - lowest_end_commit_id;
+        const CommitID max_commit_id_diff =
+            table->max_chunk_size() * static_cast<CommitID>(_DELETE_THRESHOLD_COMMIT_DIFF_FACTOR);
+
+        // Evaluate metrics
+        const bool criterion1 = _DELETE_THRESHOLD_RATE_INVALIDATED_ROWS <= invalidated_rows_ratio;
+        const bool criterion2 = max_commit_id_diff <= commit_id_diff;
+
+        if (criterion1 && criterion2) {
+          const bool success = _delete_chunk_logically(table_name, chunk_id);
+
+          if (success) {
+            DebugAssert(StorageManager::get().get_table(table_name)->get_chunk(chunk_id)->get_cleanup_commit_id() !=
+                            MvccData::MAX_COMMIT_ID,
+                        "Chunk needs to be deleted logically before deleting it physically.");
+
+            std::unique_lock<std::mutex> lock(_mutex_physical_delete_queue);
+            _physical_delete_queue.emplace(table_name, chunk_id);
+          } else {
+            std::cout << "Logical delete of chunk " << chunk_id << " failed because of MVCC conflict. Retrying..."
+                      << std::endl;
+          }
+        }
+      }
+    }  // for each chunk
+  }    // for each table
+}
+
+/**
+ * This function processes the physical-delete-queue until its empty.
+ */
+void MvccDeletePlugin::_physical_delete_loop() {
+  std::unique_lock<std::mutex> lock(_mutex_physical_delete_queue);
+
+  ChunkSpecifier chunk_spec = _physical_delete_queue.front();
+  const auto& table = StorageManager::get().get_table(chunk_spec.table_name);
+  const auto& chunk = table->get_chunk(chunk_spec.chunk_id);
+
+  DebugAssert(chunk != nullptr, "Chunk does not exist. Physical Delete can not be applied.");
+
+  // Check whether there are still active transactions that might use the chunk
+  CommitID cleanup_commit_id = chunk->get_cleanup_commit_id();
+  CommitID lowest_snapshot_commit_id = TransactionManager::get().get_lowest_active_snapshot_commit_id();
+
+  if (cleanup_commit_id < lowest_snapshot_commit_id) {
+    _delete_chunk_physically(chunk_spec.table_name, chunk_spec.chunk_id);
+    _physical_delete_queue.pop();
+  }
+}
+
+bool MvccDeletePlugin::_delete_chunk_logically(const std::string& table_name, const ChunkID chunk_id) {
+  const auto& table = StorageManager::get().get_table(table_name);
+  const auto& chunk = table->get_chunk(chunk_id);
+
+  DebugAssert(chunk != nullptr, "Chunk does not exist. Physical Delete can not be applied.");
```

Comment:
These two asserts could point to serious problems (the plugin running more than once, someone else tinkering with the chunks). I'd make them `Asserts` so they are included in release builds as well.